### ISSUE TITLE
Clean Shutdown of Servo

### ISSUE THREAD

**Issue by gterzian** [servo/servo#30849]
Requires further investigation, but some initial notes:

1. `Pipeline` doesn't wait on an ACK from script [when sending the `ExitPipeline` msg](https://github.com/servo/servo/blob/8713954a533da7f4e7ec0877b9385908596b5230/components/constellation/pipeline.rs#L409).
2. Neither does an `EventLoop` wait on an ACK from script [when sending the `ExitScriptThread` msg](https://github.com/servo/servo/blob/8713954a533da7f4e7ec0877b9385908596b5230/components/constellation/event_loop.rs#L25)
3. There is some [workflow with the BHM](https://github.com/servo/servo/blob/8713954a533da7f4e7ec0877b9385908596b5230/components/constellation/constellation.rs#L2653) involving a `closing` boolean, which is useful if the script-thread is blocking inside a script execution, however it's not good enough to ensure clean shutdown, because a script-thread could easily be already handling a message by the time it is set. The constellation does wait on an ACK from the BHM, which implies that the closing flag has been set, however that doesn't prevent script from handling one last message if past the point of checking the flag(it may be that the constellation [sends it's reply to the IPC](https://github.com/servo/servo/blob/8713954a533da7f4e7ec0877b9385908596b5230/components/constellation/constellation.rs#L1850), and then shutdowns before script is scheduled to receive it). 

Any way, requires more investigation, but looking at the panic in https://github.com/servo/servo/issues/24066 it seems clear that the problem is not that the IPC returns `None`, it's that it returns an error because the sender has already been dropped(such as `Io(Custom { kind: ConnectionReset, error: "No senders exist for this port." })` or in `Disconnected` in  https://github.com/servo/servo/issues/27146). 

The last issue linked above https://github.com/servo/servo/issues/22154 appears unrelated actually(a comment in the issue seems to correctly identify a separate problem with trying to send a message layout after it has shutdown).

_Originally posted by @gterzian in https://github.com/servo/servo/issues/27171#issuecomment-1734105018_

**Comment by mrobinson** [comment]
A common problem that we have faced is that when a task panics, other tasks wait forever for a reply to an IPC message. It seems that clean shutdown should also include cases where tasks disappear.

**Comment by gterzian** [comment]
> A common problem that we have faced is that when a task panics, other tasks wait forever for a reply to an IPC message.

Right. That should be handled over at https://github.com/servo/ipc-channel/blob/25040c93497e5cd4e2e81b997cd014015fa40fea/src/router.rs#L199

Where the callback should be called when the channel is closed, so that this can be handled in some way.

**Comment by gterzian** [comment]
TODO: 

 - https://github.com/servo/servo/issues/24850
 - https://github.com/servo/servo/issues/34887
 - https://github.com/servo/servo/issues/36747
 - https://github.com/servo/servo/issues/36848
 - https://github.com/servo/servo/issues/36845

**Comment by gterzian** [comment]
Here is a screenshot of the threads still running when the constellation is about to exit:

<img width="552" height="461" alt="Image" src="https://github.com/user-attachments/assets/44bab402-60d9-412c-8098-1f709a430057" />

And here is another one from right  before [deinit](https://github.com/servo/servo/blob/90822172726a7757ce5a76d24b496dc2273a83fe/ports/servoshell/desktop/cli.rs#L41):

<img width="507" height="258" alt="Image" src="https://github.com/user-attachments/assets/a9936bcf-e8e3-4ccb-8e18-e7e8e78bbcaa" />

Main targets for further clean shutdown, on MacOS at least:
- GStreamer(https://github.com/servo/media/issues/447)
- a loose ipc router thread -> https://github.com/servo/ipc-channel/pull/413
- gilrs(gamepad): there doesn't seem to be a way to shut it down. see https://docs.rs/crate/gilrs-core/latest/source/src/platform/macos/gamepad.rs#52
- stylo thread pools -> https://github.com/servo/servo/pull/38924

Looks like some of the other stuff that shows up above is shutdown by the time we de-init(could still look into making sure some of the shutdown before the constellation, like the JSHelper, the timer, and the profiler).

**Comment by wusyong** [comment]
> gilrs(gamepad): there doesn't seem to be a way to shut it down. see https://docs.rs/crate/gilrs-core/latest/source/src/platform/macos/gamepad.rs#52

I think for long term solution we would need to open and resolve the issue upstream: https://gitlab.com/gilrs-project/gilrs
In the meantime, we could spawn several thread lazily if they are not used. I don't think we need to spawn a thread for gamepad every time.

**Comment by gterzian** [comment]
Last one for this is at https://github.com/servo/servo/pull/39889

**Comment by Taym95** [comment]
@gterzian should close this one or keep it for future references?

**Comment by gterzian** [comment]
I think we can close as all addressable parts have now been completed.
